V8: Made it possible to assign a composite event to a specific state.

Kokenberg opened the pull request:
This results in a cleaner graph when visualizing. Also, when visualizing the graph the edge between the state and composite event is skipped.

Originally the graph would be generated as:
| Graph | Visual |
| -------|---------|
|digraph G {<br>0 [shape=ellipse, label="Initial"];<br>1 [shape=ellipse, label="Waiting"];<br>2 [shape=ellipse, label="Final"];<br>3 [shape=rectangle, label="Start"];<br>4 [shape=rectangle, label="First"];<br>5 [shape=rectangle, label="Third"];<br>6 [shape=rectangle, label="Second"];<br>0 -> 3;<br>1 -> 4;<br>1 -> 6;<br>1 -> 5;<br>2 -> 4;<br>2 -> 6;<br>3 -> 1;<br>4 -> 5;<br>5 -> 2;<br>6 -> 5;<br>}|![image](https://user-images.githubusercontent.com/15146247/151675294-754eb586-2840-4e76-b2ef-3619d58b8b3d.png)|

Skipping the edge from the state to the composite event (in this case the arrow from 'Waiting' to 'Third') results in:

| Graph | Visual |
| -------|---------|
|digraph G {<br>0 [shape=ellipse, label="Initial"];<br>1 [shape=ellipse, label="Waiting"];<br>2 [shape=ellipse, label="Final"];<br>3 [shape=rectangle, label="Start"];<br>4 [shape=rectangle, label="First"];<br>5 [shape=rectangle, label="Third"];<br>6 [shape=rectangle, label="Second"];<br>0 -> 3;<br>1 -> 4;<br>1 -> 6;<br>2 -> 4;<br>2 -> 6;<br>3 -> 1;<br>4 -> 5;<br>5 -> 2;<br>6 -> 5;<br>}|![image](https://user-images.githubusercontent.com/15146247/151675397-32e05de5-cadc-4a70-bc00-ff3eb1dbc02e.png)|

When assigning the composite event 'Third' explicitly to the 'Waiting' state results in:

| Graph | Visual |
| -------|---------|
|digraph G {<br>0 [shape=ellipse, label="Initial"];<br>1 [shape=ellipse, label="Waiting"];<br>2 [shape=ellipse, label="Final"];<br>3 [shape=rectangle, label="Start"];<br>4 [shape=rectangle, label="First"];<br>5 [shape=rectangle, label="Third"];<br>6 [shape=rectangle, label="Second"];<br>0 -> 3;<br>1 -> 4;<br>1 -> 6;<br>3 -> 1;<br>4 -> 5;<br>5 -> 2;<br>6 -> 5;<br>}|![image](https://user-images.githubusercontent.com/15146247/151675460-67c90dca-c279-47fd-a4fd-de803caaad22.png)|

Note that when assigning a composite event to a state that it removes a number of 'weird' edges. In this case the edges between the Final and First and Second events. These transitions are not defined in the state machine itself but are added due to the way the composite event is registered.

The last image visualizes the state machine best as far as I am concerned. Only when the 'First' and 'Second' events have been handled will the 'Third' event be fired and will the state machine change state from 'Waiting' to 'Final'.

```C#
public TestStateMachine(bool specificallyAssignedToWaiting)
{
    Initially(
        When(Start)
            .TransitionTo(Waiting));

    During(Waiting,
        When(First)
            .Then(context =>
            {
                context.Instance.First = true;
                context.Instance.CalledAfterAll = false;
            }),
        When(Second)
            .Then(context =>
            {
                context.Instance.SecondFirst = !context.Instance.First;
                context.Instance.Second = true;
                context.Instance.CalledAfterAll = false;
            }),
        When(Third, context => context.Instance.SecondFirst)
            .Then(context =>
            {
                context.Instance.Called = true;
                context.Instance.CalledAfterAll = true;
            })
            .Finalize()
        );

    if (specificallyAssignedToWaiting)
        // Assigning the 'Third' event specifically to the 'Waiting' state
        CompositeEvent(() => Third, x => Equals(x, Waiting), x => x.CompositeStatus, First, Second);
    else
        CompositeEvent(() => Third, x => x.CompositeStatus, First, Second);
}
```

Kokenberg commented:
@phatboyg, I found some time today to re-do my PR based on the V8 branches. Only a couple of files have been changed now. In the other PR you mentioned _NextEvents_. I never realized that that method existed. That will help me a lot in the scenario I am currently working on and which started code change in the first place. Because the 'IsComposite' flag is part of the 'Event' interface I can simple filter out composite events when reporting back to the UI.

phatboyg commented:
This is awesome, thank you. I'm really hoping that by having everything integrated into a single project with v8 that the improved code cohesion will make it easier for contributors to add great features like this without the complexity of so many different solutions and repositories. 

For me it makes it much easier to maintain, and allows for cleaner integration points since the types are aligned from the bottom up. 

Will check it closer tonight/tomorrow and get it merged. Thanks!

Kokenberg commented:
@phatboyg, I really liked poking around in the state machine code. Let me know if you think I can do anything else related to the state machine. I might be able to have a crack on it.

phatboyg commented:
So I'm looking through the syntax, and the `x => Equals(x, Waiting)` has me wondering if there might be a cleaner way to figure out the states in which the composite event is supported.

What I mean is, if the dependent events are only accepted in Waiting, I wonder if it would be possible through the data (similar to nextEvents, but akin to "acceptedstates" to see which states an event is accepted in, and then use that data when rendering the graph to know that Third can only be triggered in Waiting since First and Second are only accepted in waiting).

See what I mean? Since this seem to purely be around cleaning up the graph, it's more metadata for the graph vs. changing the behavior of states/events.

Make sense?

Kokenberg commented:
Don't forget that in the original code when asking for the nextevents on the Final state you would get back 'First' and 'Second'.
![TwoEventsInFinalState-original-code](https://user-images.githubusercontent.com/15146247/151807068-52e604fa-cf4a-4fa9-9a0f-1232cbee9d90.png)

I agree that the registration could be cleaner.

```
CompositeEvent*() => Third, Waiting, x => x.CompositeStatus, First, Second);
```
or
```
CompositeEventOn(Waiting, () => Third, x => x.CompositeStatus, First, Second);
```

I see you dislike doing the registration through the 'When' method. Any reason why?

phatboyg commented:
Because it's confusing, all events are declared via `Event` or `CompositeEvent`, making it "implicit" in a when statement is confusing.

Kokenberg commented:
Would it be allowed to pass the statecache and/or event ache by reference? That way a single point of truth can be created and an update to the cache (like setting the IsComposite flag)  can be done in a single place. That would making setting the CompositeEvent to the state based on the provided states easier as well.

phatboyg commented:
I'm not sure I understand what you mean by passing it via reference? What are you running into toying with various implementations?

Kokenberg commented:
Currently an event is passed to an activity binder. In the CompositeEvent method for example I have to update both the event in the eventcache and in the state. Just thinking if we pass the Event as ref if that would help.

Kokenberg commented:
@phatboyg, I've made some changes to the code. The overloaded _CompositeEvent_ methods I added have been removed. I've updated the logic in _CompositeEvent_ to search for states needed:

```
var states = _stateCache
	.Where(kvp =>
		kvp.Value.Events.Contains(@event) ||
		kvp.Value.Events.Any(x => events.Contains(x)) ||
		options.HasFlag(CompositeEventOptions.IncludeInitial) && Equals(kvp.Value, Initial))
	.Select(t => t.Value)
	.ToList();

foreach (State<TInstance> state in states)
{
	// Set the IsComposite flag just to make sure it is really set.
	var currentEvent = state.Events.FirstOrDefault(x => Equals(x, @event));
	if (currentEvent != null)
		currentEvent.IsComposite = true;

	During(state,
		When(events[i])
			.Execute(activity));
}
```

Only downside now is that the CompositeEvent call when creating the state machine **must** placed at the end, after all states and events have been added.

Kokenberg commented:
@phatboyg, I've updated the code, I've updated the code.
Within the 'CompositeEvent' code the 'CompositeEvent' is now assigned to all relevant states. I've removed my own overloaded methods since they are no longer needed.

```
bool Filter(State<TInstance> state)
{
    return state.Events.Contains(@event) ||
	    state.Events.Any(evt => events.Contains(evt)) ||
	    options.HasFlag(CompositeEventOptions.IncludeInitial) && Equals(state, Initial);
}

foreach (State<TInstance> state in _stateCache.Values.Where(Filter))
{
	// Set the IsComposite flag just to make sure it is really set.
	var currentEvent = state.Events.FirstOrDefault(x => Equals(x, @event));
	if (currentEvent != null)
		currentEvent.IsComposite = true;

	During(state,
		When(events[i])
			.Execute(activity));
}
```
The only downside is that the 'CompositeEvent' method **must** be called at the end of the state machine creation (or at least after the last 'state' declaration in which the CompositeEvent is used.

```
public TestStateMachine()
{
	InstanceState(x => x.CurrentState);

	Event(() => Second, x => x.CorrelateById(m => m.Message.CorrelationId));

	Initially(
		When(Start)
			.TransitionTo(Waiting));

	During(Waiting,
		When(First)
			.TransitionTo(WaitingForSecond));

	During(WaitingForSecond,
		When(Third)
			.Publish(context => new CompleteMessage(context.Instance.CorrelationId))
			.Finalize());

	CompositeEvent(() => Third, x => x.CompositeStatus, First, Second);
}
```

phatboyg commented:
I think that limitation is fine, since that's the preferred approach anyway:

![image](https://user-images.githubusercontent.com/20041/151869917-d7b41345-6a90-4103-a757-d915527157f4.png)

(from the [docs](https://masstransit-project.com/usage/sagas/automatonymous.html#composite-event))

Kokenberg commented:
Yes, I know, I've been doing that as well in my own code. However, I found tests failing because the CompositeEvent was called before anything else. I've updated those tests so they do not fail with my code changes.

phatboyg commented:
Okay, so that's a concern. Why would the tests fail? Because there is no way we can break existing state machines with this new requirement.

Kokenberg commented:
They fail because in the old way of assigning we would blindly assign the composite event to all states (excluding Initial, unless specified). Now we try and assign only to the states to which is component events are assigned. However, if we call the CompositeEvent before the other events have been assigned, no assigned state can be found.

phatboyg commented on tests/MassTransit.Tests/SagaStateMachineTests/CompositeEvent_Specs.cs:
```diff
@@ -82,12 +82,12 @@ public TestStateMachine()
-                CompositeEvent(() => Third, x => x.CompositeStatus, First, Second);
-
                 During(WaitingForSecond,
                     When(Third)
```
The issue here is that the event is now being used before it's declared...

Kokenberg replied:
True. However, the issue is that when we are looking for the states, **WaitingForSecond** has not yet been registered and can thus not be found when **CompositeEvent** is called.
![image](https://user-images.githubusercontent.com/15146247/151873425-e2c3284e-60f0-449a-8e00-104fe917f9a8.png)

Kokenberg replied:
The _stateCache does contain the **WaitingForSecond** state, but it has no events associated yet.

phatboyg replied:
It should be, all the states are initialized at startup in the constructor.

Kokenberg replied:
Yes, that is correct. Both the __stateCache_ and __eventCache_ are initialized during startup. However, the states in the __stateCache_ have no events associated yet. That is done when 'walking' the statemachine initialization.

Kokenberg replied:
@phatboyg, I think I have found a solution. I am going to investigate a bit more tomorrow. It 10:55pm, time to sign off for today.

Kokenberg replied:
Small update, I have all but one unittest working with the new setup. In the new setup CompositeEvents will have to be defined **before** Any _During_ statement. Instead of calling _During_ in the _CompositeEvent_ method I'm storing the bindings in a temporary dictionary which gets applied during _BindActivitiesToState_. So far, so good...

Kokenberg commented:
@phatboyg just checked in my solution. It comes down to 'delaying' the binding of the CompositeEvent. Now, as long as the CompositeEvent is defined _before_ it used in a _during_ the binding will be performed in the _BindActivitiesToState_ method, just like 'normal' Events are bound to the state there.

phatboyg commented:
Rebasing, and preparing to merge. I did make a few changes, but the tests all pass so it's good. Thanks!

phatboyg commented:
Done, thanks!

phatboyg commented:
Any idea why this broke the NHibernate test?

https://github.com/MassTransit/MassTransit/runs/5037521009?check_suite_focus=true

phatboyg commented:
Great, thanks. I'm sure it's something simple.

phatboyg commented:
From debugging, it looks as though the composite event itself isn't being added to the Final state (and perhaps not the initial state). The same test works if you add an interim state (not Final) and instead of calling Finalize, transitionTo(Dead) or something.

phatboyg commented:
So this is because previously, when you declared the composite event, it added itself to all of the states, putting it at the beginning of the activity list (prior to the Finalize). So the instance would still be in that state, and would be properly executed. Now that all composite event activities are added at the end of the activity list, the state has been transitioned and therefore not supported?

phatboyg commented:
Okay, the reality is, I think it's working right. Since DuringAny() doesn't include Initial/Final by default, that's why it's failing.
